src/FadingCounter.cc

```diff
@@ -14,63 +14,53 @@
 
 #include <cmath>
 
-FadingCounter::FadingCounter(): horizon(-1), precision(10), delta(-1),
-    lastTime(0), total(0)
+void
+FadingCounter::clear()
 {
-    counters.reserve(precision);
-    while (counters.size() < static_cast<unsigned int>(precision))
-        counters.push_back(0);
-}
-
-void FadingCounter::clear()
-{
-    for (int i = 0; i < precision; ++i)
-        counters[i] = 0;
+    std::fill(counters.begin(), counters.end(), 0);
     lastTime = current_dtime;
     total = 0;
 }
 
-void FadingCounter::configure(double newHorizon)
+void
+FadingCounter::configure(const time_t newHorizon)
 {
-    if (fabs(newHorizon - horizon) >= 1e-3) { // diff exceeds one millisecond
+    if (newHorizon != horizon_) {
         clear(); // for simplicity
-        horizon = newHorizon;
-        delta = horizon / precision; // may become negative or zero
+        horizon_ = newHorizon;
+        delta = horizon_ / counters.size(); // may become zero
     }
 }
 
-int FadingCounter::count(int howMany)
+uint64_t
+FadingCounter::count(uint64_t howMany)
 {
-    Must(howMany >= 0);
-
-    if (delta < 0)
+    if (horizon() < 0)
         return total += howMany; // forget nothing
 
-    if (horizon < 1e-3) // (e.g., zero)
+    if (horizon() == 0)
         return howMany; // remember nothing
 
     const double deltas = (current_dtime - lastTime) / delta;
-    if (deltas >= precision || current_dtime < lastTime) {
+    if (deltas >= counters.size() || current_dtime < lastTime) {
         clear(); // forget all values
     } else {
         // forget stale values, if any
         // fmod() or "current_dtime/delta" will overflow int for small deltas
-        const int lastSlot = static_cast<int>(fmod(lastTime, horizon) / delta);
+        const auto lastSlot = static_cast<int>(fmod(lastTime, horizon()) / delta);
         const int staleSlots = static_cast<int>(deltas);
         for (int i = 0, s = lastSlot + 1; i < staleSlots; ++i, ++s) {
-            const int idx = s % precision;
+            const auto idx = s % counters.size();
             total -= counters[idx];
             counters[idx] = 0;
-            Must(total >= 0);
```

Comment:
This assertion can be preserved if it is moved before `total` is decremented and adjusted to look like `Assure(total >= counters[idx])`.

Reply:
I removed this and the one below because they seem to exist purely for paranoia reasons and do not appear to be necessary any longer.